### ISSUE TITLE
On next, staging (in my "dot-emacs" repository) is extremely slow

### ISSUE THREAD

**Issue by jwiegley** [magit/magit#1772]
Maybe because I'm also doing a multi-core build at the same time, but staging a single file takes about 6-10 seconds.

**Comment by tarsius** [comment]
Are there many submodules and did you read https://github.com/magit/magit/wiki/FAQ#what-can-i-do-about-poor-performance?

**Comment by tarsius** [comment]
Performance still has to be improved a lot, but that will require big changes which have to wait until after the release. Did following these instructions help?

**Comment by tarsius** [comment]
(There's one thing I will attempt before the release, so there is some hope).

**Comment by jwiegley** [comment]
I haven't yet dug through it.  On my Linux machine I have a medium sized project where the very first `s` to stage a single file still takes up to 10 seconds, but thereafter is fine.  It uses no sub-modules.

**Comment by jwiegley** [comment]
It's odd how consistent this behavior is.  On my work machine, the first 's' of the day to stage a file is *extremely* slow (up to 20 seconds), but thereafter it's extremely fast.  It's like some kind of cache needing to be primed or something.

**Comment by daveloyall** [comment]
@jwiegley "Emacs starts quicker with autoloaded functions, since their libraries are not loaded right away; but you need to wait a moment when you first use such a function, while its containing file is evaluated." from https://www.gnu.org/software/emacs/manual/html_node/eintr/Autoload.html

next received some commits related to autoloads recently.

Could what you experience be related to this?

**Comment by jwiegley** [comment]
I don't think that's it, but you never know.  I'll run `elp` on it tomorrow.

**Comment by jwiegley** [comment]
@tarsius Here's something interesting: I'm staging a single file, but `magit-git-insert` is called 26 times.  I wonder if this is because I enabled `magit-backup`?

**Comment by phil-s** [comment]
Perhaps not relevant for jwiegley, but FWIW magit-auto-revert-mode is the biggest culprit for me when magit actions are slow. In my current repo, without auto-revert, staging or unstaging takes ~0.4 seconds. With auto-revert processing 72 buffers I get this:

magit-stage                              1           2.70008108    2.70008108
magit-run-git                            1           2.6999679370  2.6999679370
magit-refresh                            1           2.637550241   2.637550241
magit-revert-buffers                     1           2.277339453   2.277339453
magit-revert-buffer                      72          1.9534885929  0.0271317860

So if magit seems slow, make sure you check whether this is the issue.

**Comment by phil-s** [comment]
tarsius, when you stage or unstage something, does magit have easy access to the set of affected files? Or at least does it know if only a single file was involved?

It occurs to me that (in my above example) magit spending two seconds checking all 72 of those buffers for possible changes after I stage or unstage part of one single file is both super inefficient, and *possibly* the most annoying instance of slowness in the program at present.

If magit knew that only a single file was affected, and auto-reverted only that buffer, this would be nice and speedy.

This is probably a specific case of a more general question about the auto-revert functionality; but if this (rather annoying) specific case is also a low-hanging fruit which could be picked, it might well be worth considering.

**Comment by phil-s** [comment]
Or maybe we can reliably refrain from triggering the auto-revert behaviour *at all* for staging and unstaging operations?

Offhand I'm not seeing why that's necessary; although I do appreciate that applying the brute-force auto-revert hammer in all situations is a simple way to be certain of being in the correct state, and not doing so is a potential can of worms.

**Comment by tarsius** [comment]
@jwiegley Yeah, `magit-backup-mode` is to slow - I think I mentioned that. Fun fact, when such a feature was first suggested I rejected it on the grounds of it being to slow :-) Anyway this can be improved by not using stashes but two "parallel" wip refs - one for the index and one for the worktree. But that has to wait until after the release.

**Comment by jwiegley** [comment]
@tarsius Thank you!  But ever since I disabled `magit-backup-mode`, I haven't had any issues with speed.
